File: frontend/src/lib/storyboardSplit.ts
/**
 * Semantic storyboard packing for Quick Create.
 *
 * Product rule: scene count is driven by copy meaning, never by equal character slicing.
 * See docs/plans/2026-08-10-semantic-storyboard-count-plan.md
 */

export type DraftSplitType = "auto" | "paragraph" | "line" | "sentence";
export type StoryboardDensity = "sparse" | "standard" | "dense";

export const STORYBOARD_SCENE_MIN = 1;
export const STORYBOARD_SCENE_MAX = 100;

/** Minimum length of each side when soft-splitting on pause punctuation. */
const SOFT_EXPAND_MIN_PART = 4;
export const AUTO_MAX_CHARS_PER_SEGMENT = 52;
export const AUTO_MIN_CHARS_PER_SEGMENT = 12;
const RHYTHM_TARGET_CHARS = 36;
const RHYTHM_MAX_CHARS = 42;

export const DENSITY_CHARS_PER_SCENE: Record<StoryboardDensity, number> = {
  sparse: 60,
  standard: 40,
  dense: 28,
};

export function charsPerSceneForDensity(density: StoryboardDensity = "standard"): number {
  return DENSITY_CHARS_PER_SCENE[density] || DENSITY_CHARS_PER_SCENE.standard;
}

export const clampSceneCount = (value: number, min = STORYBOARD_SCENE_MIN, max = STORYBOARD_SCENE_MAX) => {
  if (!Number.isFinite(value)) return min;
  return Math.min(max, Math.max(min, Math.round(value)));
};

/**
 * Split draft text into semantic units using the user-selected rule.
 * Mirrors the previous QuickCreate.splitDraftByCurrentRule behavior.
 */
export function splitDraftByRule(text: string, splitType: DraftSplitType): string[] {
  const trimmed = String(text || "").trim();
  if (!trimmed) return [];

  if (splitType === "auto") return autoSplitDraft(trimmed);

  if (splitType === "paragraph") {
    return trimmed
      .split(/\n\s*\n/)
      .map((segment) => segment.trim())
      .filter(Boolean);
  }

  if (splitType === "sentence") return splitSentenceUnits(trimmed);

  // Explicit line mode
  return trimmed
    .split(/\r?\n/)
    .map((segment) => segment.trim())
    .filter(Boolean);
}

const meaningfulLength = (value: string) => value.replace(/\s+/g, "").length;
const semanticBreak = /(?:与此同时|这意味着|从而|并且|同时|而且|但是|不过|因此|所以|如果|虽然|随后|后来|其实|最终|此时|此后|只要|因为|不仅|以及|还是|并|而|但|却|让|把|将|进入|开始|最后|其中)/gu;

function splitSentenceUnits(text: string): string[] {
  const units: string[] = [];
  let buffer = "";
  const source = String(text || "");
  for (let index = 0; index < source.length; index += 1) {
    const char = source[index] || "";
    if (char === "\n") {
      if (buffer.trim()) units.push(buffer.trim());
      buffer = "";
      continue;
    }
    buffer += char;
    const next = source[index + 1] || "";
    if ("。！？!?…".includes(char)) {
      if ("。！？!?…".includes(next)) continue;
      units.push(buffer.trim());
      buffer = "";
    } else if (char === ".") {
      // Do not split decimal numbers, dates, domains, or version strings.
      if (/\d/.test(next) || next === "." || /[A-Za-z]/.test(next)) continue;
      units.push(buffer.trim());
      buffer = "";
    }
  }
  if (buffer.trim()) units.push(buffer.trim());
  return units.filter(Boolean);
}

function splitLongSentence(sentence: string): string[] {
  if (meaningfulLength(sentence) <= RHYTHM_MAX_CHARS) return [sentence.trim()];

  const tokens = sentence.split(/([，,；;：:]+)/).filter(Boolean);
  const parts: string[] = [];
  let buffer = "";
  for (const token of tokens) {
    buffer = `${buffer}${token}`.trim();
    if (/^[，,；;：:]+$/.test(token)) {
      parts.push(buffer);
      buffer = "";
    }
  }
  if (buffer) parts.push(buffer);

  if (parts.length <= 1) return splitByRhythm(sentence);

  const groups: string[] = [];
  buffer = "";
  for (const part of parts) {
    const candidate = `${buffer}${part}`.trim();
    if (buffer && meaningfulLength(candidate) > RHYTHM_MAX_CHARS) {
      groups.push(buffer);
      buffer = part;
    } else {
      buffer = candidate;
    }
  }
  if (buffer) groups.push(buffer);

  if (groups.length > 1 && meaningfulLength(groups.at(-1) || "") < AUTO_MIN_CHARS_PER_SEGMENT) {
    groups[groups.length - 2] = `${groups[groups.length - 2]}${groups[groups.length - 1]}`;
    groups.pop();
  }
  return groups.some((group) => meaningfulLength(group) > RHYTHM_MAX_CHARS)
    ? splitByRhythm(sentence)
    : (groups.length ? groups : [sentence.trim()]);
}

function splitByRhythm(text: string): string[] {
  const source = String(text || "").trim();
  if (meaningfulLength(source) <= RHYTHM_MAX_CHARS) return source ? [source] : [];

  const result: string[] = [];
  let cursor = 0;
  while (meaningfulLength(source.slice(cursor)) > RHYTHM_MAX_CHARS) {
    const limit = Math.min(source.length, cursor + RHYTHM_MAX_CHARS);
    const candidates: number[] = [];
    for (let index = cursor; index < limit; index += 1) {
      if (/[，,；;：:。！？!?…]/.test(source[index] || "") && source.length - index - 1 >= AUTO_MIN_CHARS_PER_SEGMENT) {
        candidates.push(index + 1);
      }
    }
    semanticBreak.lastIndex = cursor + AUTO_MIN_CHARS_PER_SEGMENT;
    let match: RegExpExecArray | null;
    while ((match = semanticBreak.exec(source))) {
      const position = match.index;
      if (position > limit) break;
      if (position >= limit || source.length - position < AUTO_MIN_CHARS_PER_SEGMENT) continue;
      const previous = source[position - 1] || "";
      const first = source[position] || "";
      if (/[A-Za-z0-9]/.test(previous) && /[A-Za-z0-9]/.test(first)) continue;
      candidates.push(position);
    }

    let cut = candidates.length
      ? candidates.reduce((best, position) =>
        Math.abs(position - cursor - RHYTHM_TARGET_CHARS) < Math.abs(best - cursor - RHYTHM_TARGET_CHARS)
          ? position
          : best,
      candidates[0]!)
      // ponytail: fixed-width CJK fallback; replace with a tokenizer only if cut quality is measured as a problem.
      : Math.min(cursor + RHYTHM_TARGET_CHARS, source.length - Math.max(AUTO_MIN_CHARS_PER_SEGMENT, source.length - cursor - RHYTHM_TARGET_CHARS));
    cut = Math.max(cursor + AUTO_MIN_CHARS_PER_SEGMENT, Math.min(cut, limit));
    if (cut > 0 && cut < source.length && /[A-Za-z0-9]/.test(source[cut - 1] || "") && /[A-Za-z0-9]/.test(source[cut] || "")) {
      for (let candidate = cut - 1; candidate >= cursor + AUTO_MIN_CHARS_PER_SEGMENT; candidate -= 1) {
        if (!(/[A-Za-z0-9]/.test(source[candidate - 1] || "") && /[A-Za-z0-9]/.test(source[candidate] || ""))) {
          cut = candidate;
          break;
        }
      }
    }
    if (cut <= cursor) cut = Math.min(limit, cursor + RHYTHM_TARGET_CHARS);
    result.push(source.slice(cursor, cut).trim());
    cursor = cut;
  }
  const tail = source.slice(cursor).trim();
  if (tail) result.push(tail);
  return result.length ? result : [source];
}

/** Split only at sentence or explicit pause boundaries; never character-slice. */
export function autoSplitDraft(text: string): string[] {
  const trimmed = String(text || "").trim();
  if (!trimmed) return [];
  const sentences = splitSentenceUnits(trimmed);
  const units: string[] = [];
  for (const sentence of sentences) {
    const clean = sentence.trim();
    if (clean) units.push(...splitLongSentence(clean));
  }
  return units.length ? units : [trimmed];
}

/** Keep custom targets soft without allowing a packed unit over the rhythm ceiling. */
export function rebalanceLongUnits(units: string[]): string[] {
  return units.flatMap((unit) => {
    const clean = unit.trim();
    return meaningfulLength(clean) > RHYTHM_MAX_CHARS ? splitLongSentence(clean) : (clean ? [clean] : []);
  });
}

/**
 * Soft-expand units on pause punctuation (，；) only when both sides are substantial.
 * Never hard-cuts mid-phrase without punctuation.
 * Keeps the pause mark on the *left* segment so clause boundaries stay speakable.
 */
export function softExpandByPause(units: string[]): string[] {
  const expanded: string[] = [];
  for (const unit of units) {
    const trimmed = unit.trim();
    if (!trimmed) continue;
    // Keep delimiters: "A，B，C。" → ["A，", "B，", "C。"]
    const tokens = trimmed.split(/([，,；;]+)/).filter((part) => part.length > 0);
    const parts: string[] = [];
    let buffer = "";
    for (const token of tokens) {
      if (/^[，,；;]+$/.test(token)) {
        buffer = `${buffer}${token}`.trim();
        if (buffer) {
          parts.push(buffer);
          buffer = "";
        }
      } else {
        buffer = `${buffer}${token}`.trim();
      }
    }
    if (buffer) parts.push(buffer);

    if (parts.length > 1 && parts.every((part) => Array.from(part.replace(/[，,；;]/g, "")).length >= SOFT_EXPAND_MIN_PART)) {
      expanded.push(...parts);
    } else {
      expanded.push(trimmed);
    }
  }
  return expanded;
}

/**
 * Suggest storyboard count from draft semantics.
 */
export function suggestSceneCount(
  text: string,
  splitType: DraftSplitType,
  options?: { softExpand?: boolean; min?: number; max?: number },
): number {
  const softExpand = options?.softExpand !== false;
  let units = splitDraftByRule(text, splitType);
  if (softExpand) units = softExpandByPause(units);
  return clampSceneCount(units.length || 1, options?.min, options?.max);
}

/** Count CJK/meaningful characters for rhythm estimates (ignore pure whitespace). */
export function countCopyChars(text: string): number {
  return Array.from(String(text || "").replace(/\s+/g, "")).length;
}

/**
 * Rhythm-based scene count: total chars ÷ target chars per scene.
 * Default ~40 chars/scene ≈ 8–12s oral delivery at typical short-video pace.
 */
export function suggestRhythmSceneCount(
  textOrCharCount: string | number,
  options?: { charsPerScene?: number; min?: number; max?: number },
): number {
  const chars =
    typeof textOrCharCount === "number"
      ? Math.max(0, textOrCharCount)
      : countCopyChars(textOrCharCount);
  const per = Math.max(12, options?.charsPerScene ?? 40);
  return clampSceneCount(Math.max(1, Math.round(chars / per)), options?.min, options?.max);
}

export interface StoryboardRecommendation {
  semantic: number;
  rhythm: number;
  charCount: number;
  units: string[];
  longestChars: number;
  longestSeconds: number;
  warnings: string[];
  /** Preferred default adopt target: semantic first. */
  preferred: number;
  density: StoryboardDensity;
  recommendedSceneCount: number;
  actualSceneCount: number;
  targetSceneCount: number | null;
  estimatedDurationSeconds: number;
}

/**
 * Step-2 analysis after pure copy is ready (no LLM required).
 */
export function analyzeStoryboardRecommendation(
  text: string,
  splitType: DraftSplitType,
  options?: {
    softExpand?: boolean;
    charsPerScene?: number;
    density?: StoryboardDensity;
    targetSceneCount?: number | null;
    min?: number;
    max?: number;
  },
): StoryboardRecommendation {
  const charCount = countCopyChars(text);
  const semantic = suggestSceneCount(text, splitType, options);
  const density = options?.density || "standard";
  const rhythm = suggestRhythmSceneCount(charCount, {
    ...options,
    charsPerScene: options?.charsPerScene ?? charsPerSceneForDensity(density),
  });
  let units = splitDraftByRule(text, splitType);
  if (options?.softExpand !== false) units = softExpandByPause(units);
  units = rebalanceLongUnits(units);
  const longestChars = Math.max(0, ...units.map((unit) => meaningfulLength(unit)));
  const longestSeconds = (longestChars / 260) * 60;
  const warnings: string[] = [];
  if (longestChars > AUTO_MAX_CHARS_PER_SEGMENT) {
    warnings.push(`仍有 ${longestChars} 字的旁白无法在现有标点处安全拆分`);
  }
  if (longestSeconds > (AUTO_MAX_CHARS_PER_SEGMENT * 60) / 260) {
    warnings.push(`最长分镜预计 ${longestSeconds.toFixed(1)} 秒，建议启用视觉节拍或补充停顿标点`);
  }
  const targetSceneCount = options?.targetSceneCount == null
    ? null
    : clampSceneCount(options.targetSceneCount, options?.min, options?.max);
  const actualUnits = rebalanceLongUnits(
    targetSceneCount == null ? units : packSemanticUnits(units, targetSceneCount),
  );
  const actualSceneCount = actualUnits.length || 1;
  if (targetSceneCount != null && actualSceneCount !== targetSceneCount) {
    warnings.push(`目标 ${targetSceneCount} 镜，实际采用 ${actualSceneCount} 个自然语义镜头`);
  }
  return {
    semantic,
    rhythm,
    charCount,
    units,
    longestChars,
    longestSeconds,
    warnings,
    preferred: semantic,
    density,
    recommendedSceneCount: semantic,
    actualSceneCount,
    targetSceneCount,
    estimatedDurationSeconds: Math.max(1, Math.round((charCount / 260) * 60)),
  };
}

/**
 * Pack semantic units into targetCount scenes.
 * - More units than target → merge adjacent units evenly.
 * - Fewer units than target → keep units as-is (do NOT character-slice).
 */
export function packSemanticUnits(units: string[], targetCount: number): string[] {
  const clean = units.map((unit) => unit.trim()).filter(Boolean);
  if (clean.length === 0) return [];

  const target = clampSceneCount(targetCount || clean.length);
  if (clean.length === target) return clean;

  if (clean.length > target) {
    return Array.from({ length: target }, (_, index) => {
      const start = Math.floor((index * clean.length) / target);
      const end = Math.floor(((index + 1) * clean.length) / target);
      return clean.slice(start, Math.max(end, start + 1)).join("").trim();
    }).filter(Boolean);
  }

  // clean.length < target: semantic priority — never invent mid-sentence cuts.
  return clean;
}

/**
 * Full pipeline: rule split → optional soft expand → pack to target.
 */
/**
 * Merge hard mid-word cuts like 「科学家发」+「现，光速」.
 */
export function healMidCuts(segments: string[]): string[] {
  const clean = segments.map((s) => s.trim()).filter(Boolean);
  if (clean.length < 2) return clean;
  const terminalOrPause = /[。！？.!?…，,；;]$/;
  const merged: string[] = [];
  let index = 0;
  while (index < clean.length) {
    let current = clean[index]!;
    while (index + 1 < clean.length) {
      const next = clean[index + 1]!;
      if (terminalOrPause.test(current)) break;
      const leftEnd = current[current.length - 1] || "";
      const rightStart = next[0] || "";
      const cjk =
        /[\u4e00-\u9fff]/.test(leftEnd) && /[\u4e00-\u9fff]/.test(rightStart);
      const alnum = /[a-zA-Z0-9]/.test(leftEnd) && /[a-zA-Z0-9]/.test(rightStart);
      if (!cjk && !alnum) break;
      // Long independent clauses without periods — leave alone
      if (Array.from(current).length >= 24 && Array.from(next).length >= 24 && /[这那他她你我神心]/.test(rightStart)) {
        break;
      }
      const should =
        Array.from(current).length <= 40
        || Array.from(next).length <= 12
        || /[现得着过了们到上下中里出来去]/.test(rightStart)
        || Array.from(current).length < 16;
      if (!should) break;
      current = `${current}${next}`;
      index += 1;
    }
    merged.push(current);
    index += 1;
  }
  return merged;
}

export function buildStoryboardNarrations(
  text: string,
  splitType: DraftSplitType,
  targetCount: number,
  options?: { softExpand?: boolean; heal?: boolean },
): string[] {
  const softExpand = options?.softExpand !== false;
  const heal = options?.heal !== false;
  let units = splitDraftByRule(text, splitType);
  if (softExpand) units = softExpandByPause(units);
  let packed = packSemanticUnits(units, targetCount);
  packed = rebalanceLongUnits(packed);
  if (heal) packed = healMidCuts(packed);
  return packed;
}

/**
 * Detect likely mid-phrase hard cuts (heuristic for UI warnings / tests).
 * Returns true if any adjacent pair looks like a broken CJK phrase with no terminal punct.
 */
export function looksLikeMidSentenceCut(segments: string[]): boolean {
  if (segments.length < 2) return false;
  const terminal = /[。！？.!?]$/;
  for (let i = 0; i < segments.length - 1; i += 1) {
    const left = segments[i]?.trim() || "";
    const right = segments[i + 1]?.trim() || "";
    if (!left || !right) continue;
    if (terminal.test(left)) continue;
    // Very short trailing fragment on the right often means a hard cut ("己", "的自己")
    if (Array.from(right).length <= 3 && !/[，,；;]/.test(left)) return true;
    // Left ends mid-clause without pause mark and right continues same sentence style
    if (!/[，,；;。！？.!?]$/.test(left) && Array.from(left).length >= 2 && Array.from(right).length >= 2) {
      // Only flag when left is not a full clause ending with pause
      if (Array.from(left).length < 8 || Array.from(right).length < 8) {
        // weak signal; skip short noise
      }
    }
  }
  return false;
}

File: pixelle_video/utils/storyboard_split.py
"""Semantic storyboard packing — mirrors frontend/src/lib/storyboardSplit.ts.

Product rules:
- Scene count follows copy meaning, never equal character slicing.
- Never invent mid-word / mid-phrase hard cuts to hit a target count.
- Soft-expand on pause punctuation may create multi-clip sentences; callers
  using continuous TTS should prefer soft_expand=False.
"""

from __future__ import annotations

import re
from typing import Literal

DraftSplitType = Literal["auto", "paragraph", "line", "sentence"]

STORYBOARD_SCENE_MIN = 1
STORYBOARD_SCENE_MAX = 100
SOFT_EXPAND_MIN_PART = 4
AUTO_MAX_CHARS_PER_SEGMENT = 52
AUTO_MIN_CHARS_PER_SEGMENT = 12
RHYTHM_TARGET_CHARS = 36
RHYTHM_MAX_CHARS = 42

_TERMINAL = re.compile(r"[。！？.!?…]+$")
_PAUSE = re.compile(r"[，,；;]+$")
_ASCII_LETTER = re.compile(r"[A-Za-z]")
_SEMANTIC_BREAK = re.compile(
    r"(?:与此同时|这意味着|从而|并且|同时|而且|但是|不过|因此|所以|如果|虽然|随后|后来|其实|最终|此时|此后|只要|因为|不仅|以及|还是|并|而|但|却|让|把|将|进入|开始|最后|其中)"
)


def clamp_scene_count(value: int, minimum: int = STORYBOARD_SCENE_MIN, maximum: int = STORYBOARD_SCENE_MAX) -> int:
    try:
        number = int(value)
    except (TypeError, ValueError):
        number = minimum
    return min(maximum, max(minimum, number))


def split_draft_by_rule(text: str, split_type: DraftSplitType = "auto") -> list[str]:
    trimmed = str(text or "").strip()
    if not trimmed:
        return []

    if split_type == "auto":
        return auto_split_draft(trimmed)

    if split_type == "paragraph":
        return [segment.strip() for segment in re.split(r"\n\s*\n", trimmed) if segment.strip()]

    if split_type == "sentence":
        # Keep terminal punctuation with the sentence.
        return _split_sentence_units(trimmed)

    # Explicit line mode
    return [line.strip() for line in re.split(r"\r?\n", trimmed) if line.strip()]


def _meaningful_length(text: str) -> int:
    return len(re.sub(r"\s+", "", str(text or "")))


def _split_sentence_units(text: str) -> list[str]:
    """Split at sentence punctuation without breaking decimal/date/domain tokens."""
    units: list[str] = []
    buffer: list[str] = []
    source = str(text or "")
    for index, char in enumerate(source):
        if char == "\n":
            if buffer:
                units.append("".join(buffer).strip())
                buffer = []
            continue
        buffer.append(char)
        next_char = source[index + 1] if index + 1 < len(source) else ""
        if char in "。！？!?…":
            if next_char in "。！？!?…":
                continue
            units.append("".join(buffer).strip())
            buffer = []
        elif char == ".":
            # Decimal numbers, dates, domains, and version strings are not boundaries.
            if next_char.isdigit() or next_char == "." or _ASCII_LETTER.fullmatch(next_char or ""):
                continue
            units.append("".join(buffer).strip())
            buffer = []
    if buffer:
        units.append("".join(buffer).strip())
    return [unit for unit in units if unit]


def _split_long_sentence(sentence: str) -> list[str]:
    """Split long narration at pauses, semantic joins, then a bounded beat."""
    if _meaningful_length(sentence) <= RHYTHM_MAX_CHARS:
        return [sentence.strip()]

    pieces = [piece for piece in re.split(r"([，,；;：:]+)", sentence) if piece]
    parts: list[str] = []
    buffer = ""
    for piece in pieces:
        buffer = f"{buffer}{piece}".strip()
        if re.fullmatch(r"[，,；;：:]+", piece):
            parts.append(buffer)
            buffer = ""
    if buffer:
        parts.append(buffer)

    if len(parts) <= 1:
        return _split_by_rhythm(sentence)

    groups: list[str] = []
    buffer = ""
    for part in parts:
        candidate = f"{buffer}{part}".strip()
        if buffer and _meaningful_length(candidate) > RHYTHM_MAX_CHARS:
            groups.append(buffer)
            buffer = part
        else:
            buffer = candidate
    if buffer:
        groups.append(buffer)

    # Avoid creating a tiny tail when the preceding clause is a safe merge.
    if len(groups) > 1 and _meaningful_length(groups[-1]) < AUTO_MIN_CHARS_PER_SEGMENT:
        groups[-2] = f"{groups[-2]}{groups[-1]}"
        groups.pop()
    if any(_meaningful_length(group) > RHYTHM_MAX_CHARS for group in groups):
        return _split_by_rhythm(sentence)
    return groups or [sentence.strip()]


def _split_by_rhythm(text: str) -> list[str]:
    """Guarantee bounded units when copy omitted punctuation.

    Connector boundaries preserve source text and usually match a natural
    visual change. The final fixed-width fallback is deliberately bounded so
    a 60-character unpunctuated sentence cannot become a 14-second scene.
    """
    source = str(text or "").strip()
    if _meaningful_length(source) <= RHYTHM_MAX_CHARS:
        return [source] if source else []

    result: list[str] = []
    cursor = 0
    while _meaningful_length(source[cursor:]) > RHYTHM_MAX_CHARS:
        limit = min(len(source), cursor + RHYTHM_MAX_CHARS)
        candidates: list[int] = []
        for match in re.finditer(r"[，,；;：:。！？!?…]+", source[cursor:limit]):
            position = cursor + match.end()
            if position < limit and len(source) - position >= AUTO_MIN_CHARS_PER_SEGMENT:
                candidates.append(position)
        for match in _SEMANTIC_BREAK.finditer(source, cursor + AUTO_MIN_CHARS_PER_SEGMENT, limit + 1):
            position = match.start()
            if position >= limit or len(source) - position < AUTO_MIN_CHARS_PER_SEGMENT:
                continue
            previous = source[position - 1] if position else ""
            first = source[position] if position < len(source) else ""
            if previous.isascii() and previous.isalnum() and first.isascii() and first.isalnum():
                continue
            candidates.append(position)

        if candidates:
            cut = min(candidates, key=lambda position: abs((position - cursor) - RHYTHM_TARGET_CHARS))
        else:
            # Keep a meaningful tail while staying below the 10-second rhythm
            # threshold used by the analysis warning.
            # ponytail: fixed-width CJK fallback; replace with a tokenizer only if cut quality is measured as a problem.
            cut = min(cursor + RHYTHM_TARGET_CHARS, len(source) - _minimum_tail_length(source, cursor))
            cut = max(cursor + AUTO_MIN_CHARS_PER_SEGMENT, min(cut, limit))
            cut = _move_off_ascii_word(source, cut, cursor + AUTO_MIN_CHARS_PER_SEGMENT, limit)

        if cut <= cursor:
            cut = min(limit, cursor + RHYTHM_TARGET_CHARS)
        result.append(source[cursor:cut].strip())
        cursor = cut

    tail = source[cursor:].strip()
    if tail:
        result.append(tail)
    return result or [source]


def _minimum_tail_length(source: str, cursor: int) -> int:
    """Return a tail size without adding another public tuning parameter."""
    return max(AUTO_MIN_CHARS_PER_SEGMENT, len(source) - cursor - RHYTHM_TARGET_CHARS)


def _move_off_ascii_word(source: str, cut: int, minimum: int, maximum: int) -> int:
    if cut <= 0 or cut >= len(source):
        return cut
    if not (source[cut - 1].isascii() and source[cut - 1].isalnum() and source[cut].isascii() and source[cut].isalnum()):
        return cut
    for candidate in range(cut - 1, minimum - 1, -1):
        if not (source[candidate - 1].isascii() and source[candidate - 1].isalnum() and source[candidate].isascii() and source[candidate].isalnum()):
            return candidate
    for candidate in range(cut + 1, maximum + 1):
        if not (source[candidate - 1].isascii() and source[candidate - 1].isalnum() and source[candidate].isascii() and source[candidate].isalnum()):
            return candidate
    return cut


def auto_split_draft(text: str) -> list[str]:
    """Create safe, duration-aware narration units without rewriting source text."""
    trimmed = str(text or "").strip()
    if not trimmed:
        return []

    sentences = _split_sentence_units(trimmed)
    units: list[str] = []
    for sentence in sentences:
        clean = sentence.strip()
        if clean:
            units.extend(_split_long_sentence(clean))
    return units or [trimmed]


def rebalance_long_units(units: list[str]) -> list[str]:
    """Re-split packed custom units without allowing a long tail to return."""
    result: list[str] = []
    for unit in units:
        clean = str(unit or "").strip()
        if not clean:
            continue
        result.extend(_split_long_sentence(clean) if _meaningful_length(clean) > RHYTHM_MAX_CHARS else [clean])
    return result


def soft_expand_by_pause(units: list[str]) -> list[str]:
    """
    Soft-expand on ，； only when both sides are substantial.

    Preserves the pause mark on the *left* segment so continuous TTS / hold
    boundaries still read as natural clause ends (not mid-word hard cuts).
    """
    expanded: list[str] = []
    for unit in units:
        trimmed = unit.strip()
        if not trimmed:
            continue
        # Split while keeping delimiters
        pieces = re.split(r"([，,；;]+)", trimmed)
        # pieces: [text, delim, text, delim, text, ...]
        parts: list[str] = []
        buf = ""
        for piece in pieces:
            if not piece:
                continue
            if re.fullmatch(r"[，,；;]+", piece):
                buf = (buf + piece).strip()
                if buf:
                    parts.append(buf)
                    buf = ""
            else:
                buf = (buf + piece).strip()
        if buf:
            parts.append(buf)

        if (
            len(parts) > 1
            and all(len(re.sub(r"\s+", "", part)) >= SOFT_EXPAND_MIN_PART for part in parts)
        ):
            expanded.extend(parts)
        else:
            expanded.append(trimmed)
    return expanded


def pack_semantic_units(units: list[str], target_count: int) -> list[str]:
    clean = [unit.strip() for unit in units if unit and unit.strip()]
    if not clean:
        return []

    target = clamp_scene_count(target_count or len(clean))
    if len(clean) == target:
        return clean

    if len(clean) > target:
        packed: list[str] = []
        for index in range(target):
            start = (index * len(clean)) // target
            end = ((index + 1) * len(clean)) // target
            chunk = "".join(clean[start : max(end, start + 1)]).strip()
            if chunk:
                packed.append(chunk)
        return packed

    # Fewer units than target: never invent mid-sentence character cuts.
    return clean


def heal_mid_cuts(segments: list[str]) -> list[str]:
    """
    Merge adjacent segments that look like hard mid-word / mid-phrase cuts.

    Example: ``科学家发`` + ``现，光速…`` → one segment.
    Does not merge when left already ends with terminal or pause punctuation.
    """
    clean = [segment.strip() for segment in segments if segment and segment.strip()]
    if len(clean) < 2:
        return clean

    merged: list[str] = []
    index = 0
    while index < len(clean):
        current = clean[index]
        while index + 1 < len(clean):
            nxt = clean[index + 1]
            if _should_merge_mid_cut(current, nxt):
                current = f"{current}{nxt}"
                index += 1
                continue
            break
        merged.append(current)
        index += 1
    return merged


def _should_merge_mid_cut(left: str, right: str) -> bool:
    left = left.strip()
    right = right.strip()
    if not left or not right:
        return False
    if _TERMINAL.search(left) or _PAUSE.search(left):
        return False
    # Left ends mid-flow without punctuation; right continues CJK/alnum.
    left_end = left[-1]
    right_start = right[0]
    cjk = "\u4e00" <= left_end <= "\u9fff" and "\u4e00" <= right_start <= "\u9fff"
    alnum = (left_end.isalnum() and right_start.isalnum() and left_end.isascii() and right_start.isascii())
    if not (cjk or alnum):
        return False
    # Avoid merging two full independent sentences that merely lack periods
    # (both long and right looks like a new sentence start with known openers).
    if len(left) >= 24 and len(right) >= 24 and right_start in "这那其他她你我神心其其":
        # Ambiguous — do not force merge long pairs
        return False
    # Short left tail without punct is almost always a hard cut ("发"+"现")
    if len(left) <= 40 or len(right) <= 12:
        return True
    # Medium: merge when right starts with continuation-like chars
    if right_start in "现得着过了们到上下中里出来去着过":
        return True
    return len(left) < 16


def build_storyboard_narrations(
    text: str,
    split_type: DraftSplitType = "auto",
    target_count: int = 5,
    *,
    soft_expand: bool = True,
    heal: bool = True,
) -> list[str]:
    units = split_draft_by_rule(text, split_type)
    if soft_expand:
        units = soft_expand_by_pause(units)
    packed = pack_semantic_units(units, target_count)
    packed = rebalance_long_units(packed)
    if heal:
        packed = heal_mid_cuts(packed)
    return packed
